### PR TITLE
Only allow constants for prepared statement parameters

### LINKED ISSUE #15297: Prepared Statements allow some non-constant expressions as parameters
ConstantExpressionVerifier is used mainly to check that the parameters passed to execute a prepared statement are constants, meaning that they don't select from any tables.  However, it right now only just checks for column references, which means it misses a few cases

1. `SELECT * FROM table` -- it doesn't recognize the 'AllColumns' Expression as being non-constant, so this expression is allowed, even though `SELECT my_col FROM table` will fail.
2. `SELECT 1 FROM table` -- even though this expression doesn't reference any columns, it's not constant, but the ConstantExpressionVerifier doesn't recognize that

It also incorrectly considers certain query shapes as being non-constant even though they only SELECT from a "values" clause.  Though in my opinion this is less important
e.g. 
1. SELECT a from (VALUES 1)t(a) is considered non-constant because it has a column reference, but SELECT 1 is fine.

Should we simply make the constant expression verifier more restrictive so that it excludes anything with an AllColumns expression (*) and anything with a Table expression?  This would still allow parameters for prepared statements to contain arbitrary expressions (e.g. complex regexes would still be fine), but would prevent any expression that accesses outside data.  Alternatively we can restrict Prepared Statements to only literals.

### PR DESCRIPTION
**Pull request by rschlussel**
We were allowing anything without an explicit column reference, which
included SELECT * from table and SELECT 1 from table.  Now only
expressions that reference neither tables nor columns (nor *) will be
allowed.

Test plan -  unit tests

Fixes #15297 

```
== RELEASE NOTES ==

General Changes
* Fix an issue where prepared statements would allow some non-constant parameters.
```

### CODE REVIEW

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 1]
```diff
@@ -4760,6 +4773,44 @@ public void testExecuteWithParametersInGroupBy()
...
+                    .addPreparedStatement("my_query", query)
+                    .build();
+            computeActual(session, "EXECUTE my_query USING \"nationkey\"");
+            fail("nonLiteral parameters should fail");
+        }
+        catch (SemanticException e) {
+            assertEquals(e.getCode(), EXPRESSION_NOT_CONSTANT);
+        }
+        catch (RuntimeException e) {
+            assertEquals(e.getMessage(), "line 1:24: Constant expression cannot contain column references");
```
Comment (kaikalur) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:4810]: Also add a test for passing * - AllColumns I assume that you just added a visitor above.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ConstantExpressionVerifier.java [hunk 2]
```diff
@@ -67,5 +69,17 @@ protected Void visitFieldReference(FieldReference node, Void context)
...
+        protected Void visitAllColumns(AllColumns node, Void context)
+        {
+            throw new SemanticException(EXPRESSION_NOT_CONSTANT, expression, "Constant expression cannot contain column references");
+        }
+
+        @Override
+        protected Void visitTable(Table node, Void context)
+        {
+            throw new SemanticException(EXPRESSION_NOT_CONSTANT, expression, "Constant expression cannot contain table references");
+        }
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ConstantExpressionVerifier.java:83]: How about things like UDFs? Maybe we should disallow those as well?

**Review by kaikalur — COMMENTED** [review]
I see there is a RowExpressionInterpeter class. Maybe you can use that to make sure the expression evaluates a constant and reject if it doesn't?

**Comment by rschlussel** [comment]
@kaikalur I updated this to only allow literal expressions.  I'm not sure if there are other cases like the map constructor though that I've missed.

**Review by kaikalur — CHANGES_REQUESTED** [review]

**Comment by kaikalur** [comment]
> @kaikalur I updated this to only allow literal expressions. I'm not sure if there are other cases like the map constructor though that I've missed.

Actually, I remember now - can't we just use RowExpressionInterpter like I suggested earier?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ParameterVerifier.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+public final class ParameterVerifier
+{
+    private ParameterVerifier() {}
+
+    public static void verifyParameterIsValid(Expression expression)
+    {
+        new ParameterVerifierVisitor().process(expression, null);
+    }
+
+    private static class ParameterVerifierVisitor
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ParameterVerifier.java:42]: Also, I see another similar class:

presto-main/src/main/java/com/facebook/presto/sql/analyzer/ConstantExpressionVerifier.java

Can we use that?

**Comment by rschlussel** [comment]
@kaikalur updated the pr.  As discussed over chat, I switched this back to allowing any constants, as we decided we are okay with any functions as long as the expressions are constant.  Also, as discussed we can't use the RowExpressionInterpreter since the expressions have not been converted to row expressions at the time we replace the prepared statements.

**Review by rongrong — APPROVED** [review]

**Pull request merged**